src/System.Private.CoreLib/shared/System/Text/Unicode/Utf8Utility.cs

```diff
@@ -44,6 +60,143 @@ public static unsafe int GetIndexOfFirstInvalidUtf8Sequence(ReadOnlySpan<byte> u
+
+                    nuint utf16CharCount = ASCIIUtility.GetIndexOfFirstNonAsciiByte(pBuffer, bufferLength);
+
+                    if (utf16CharCount != bufferLength)
+                    {
+                        // Found at least one non-ASCII byte, so fall down the slower (but still vectorized) code paths.
+                        // Given well-formed UTF-8 input, we can compute the number of resulting UTF-16 code units
+                        // using the following formula:
+                        //
+                        // utf16CharCount = utf8ByteCount - numUtf8ContinuationBytes + numUtf8FourByteHeaders
+
+                        utf16CharCount = bufferLength;
+
+                        Vector128<sbyte> vecAllC0 = Vector128.Create(unchecked((sbyte)0xC0));
+                        Vector128<sbyte> vecAll80 = Vector128.Create(unchecked((sbyte)0x80));
+                        Vector128<sbyte> vecAll6F = Vector128.Create(unchecked((sbyte)0x6F));
+
+                        {
+                            // Perform an aligned read of the first part of the buffer.
+                            // We'll mask out any data at the start of the buffer we don't care about.
+                            //
+                            // For example, if (pBuffer MOD 16) = 2:
+                            // [ AA BB CC DD ... ] <-- original vector
+                            // [ 00 00 CC DD ... ] <-- after PANDN operation
+
+                            nint offset = -((nint)pBuffer & (sizeof(Vector128<sbyte>) - 1));
+                            Vector128<sbyte> shouldBeMaskedOut = Sse2.CompareGreaterThan(Vector128.Create((byte)((int)offset + sizeof(Vector128<sbyte>) - 1)).AsSByte(), VectorOfElementIndices);
+                            Vector128<sbyte> thisVector = Sse2.AndNot(shouldBeMaskedOut, Unsafe.Read<Vector128<sbyte>>(pBuffer + offset));
+
+                            // If there's any data at the end of the buffer we don't care about, mask it out now.
+                            // If this happens the 'bufferLength' value will be a lie, but it'll cause all of the
+                            // branches later in the method to be skipped, so it's not a huge problem.
+
+                            if (bufferLength < (nuint)offset + (uint)sizeof(Vector128<sbyte>))
+                            {
+                                Vector128<sbyte> shouldBeAllowed = Sse2.CompareLessThan(VectorOfElementIndices, Vector128.Create((byte)((int)bufferLength - (int)offset)).AsSByte());
+                                thisVector = Sse2.And(shouldBeAllowed, thisVector);
+                                bufferLength = (nuint)offset + (uint)sizeof(Vector128<sbyte>);
+                            }
+
+                            uint maskOfContinuationBytes = (uint)Sse2.MoveMask(Sse2.CompareGreaterThan(vecAllC0, thisVector));
+                            uint countOfContinuationBytes = Popcnt.PopCount(maskOfContinuationBytes);
```

Comment:
`PopCount`: did you think about using `System.Numerics.BitOperations.PopCount`?
So one could use this code path with just `Sse2` available.
If popcnt is supported by the cpu, it will result in code as written.
If popcnt is not supported, there's still the benefit from sse2.

Reply:
Using `BitOperations.PopCount` is also going to make the code more AOT friendly.

Reply:
I considered it but didn't run benchmarks with the `COMPLUS_ENABLEPOPCNT=0` env var set, and initially I was concerned that performance would tank if we couldn't use the primitive, especially for small inputs. Will rerun the benchmarks with both configurations to see how things go.